### PR TITLE
[Feature] - gLiveView - Mithril Signer Section

### PR DESCRIPTION
**Pull request by asnakep**
Hello CNTools/Koios Team,

I would like to propose the mithril-signer section I added in gLiveView, it can be useful signer monitoring.

All suggestions from previous PR #1786 have been applied.

1. Moving the Registered / Registered Total row to the bottom might make for a better UX experience...
2. Should make this a choice of user input, like ENABLE_RELAY_ENDPOINT...
   The METRICS_SERVER_PORT should also be a choice of user input.
   Also METRICS_SERVER_IP from user input... 
3. Though awk is fairly quick, using bash internal REMATCH is even faster....
4. Could be moved outside of loop to not have to check this on every refresh... 
5. Its also hardcoded for cnode name, please use ${CNODE_VNAME} instead...

gLiveView with new feature has been tested on a relay too.

Thank you,
kind regards,
Manuel

## Description
mithril-signer section for gLiveView

## Updated files:
gLiveView.sh
mithril-signer.sh
mithril.library

## Where should the reviewer start?
on a Block Producer:
- run ./mithril-signer.sh -e to recreate mithril.env with ENABLE_METRICS_SERVER=true and chosen IP:PORT for endpoint
- restart mithril-signer service
- run gLiveView.sh

## Motivation and context
Section has been added to have a clear view of mithril-signer status.

Compact view
image

Verbose view
image

### CODE REVIEW

File: scripts/cnode-helper-scripts/gLiveView.sh
Comment (TrevorBenson) [inline scripts/cnode-helper-scripts/gLiveView.sh:63]: Restore the existing version. Although maybe increment to 1.30.1. I'll defer to @rdlrt or @Scitz0 on correct semver for gLiveView changes.

File: scripts/cnode-helper-scripts/gLiveView.sh
Comment (TrevorBenson) [inline scripts/cnode-helper-scripts/gLiveView.sh:406]: @asnakep I have not run into instances where `clear` might not be available as both debian and enterprise linux based distros will have the libncurses or ncurses-libs packages installed from guild-deploy.sh which should ensure `clear` is always available.

That said I'm not against relying on bash `printf` for screen clears.  Although, it might be nice to add a note to the description mentioning the change and reasoning for others to consider.

File: scripts/cnode-helper-scripts/gLiveView.sh
Comment (TrevorBenson) [inline scripts/cnode-helper-scripts/gLiveView.sh:731]: I think this is an artifact from the source branch starting from alpha before recent commits were merged. I would have thought your rebase off upstream/alpha would have dropped this change, bit it appears to be stored in the commit instead of cleared up by the rebase operation.

File: scripts/cnode-helper-scripts/gLiveView.sh
Comment (TrevorBenson) [inline scripts/cnode-helper-scripts/gLiveView.sh:750]: Ditto, I believe this is unintended.

File: scripts/cnode-helper-scripts/gLiveView.sh
Comment (TrevorBenson) [inline scripts/cnode-helper-scripts/gLiveView.sh:817]: Ditto, I believe this is unintended.

File: scripts/cnode-helper-scripts/gLiveView.sh
Comment (TrevorBenson) [inline scripts/cnode-helper-scripts/gLiveView.sh:858]: Ditto, I believe this is unintended.

File: scripts/cnode-helper-scripts/gLiveView.sh
Comment (asnakep) [reply scripts/cnode-helper-scripts/gLiveView.sh:406]: I didn't modify clrScreen function

Comment (TrevorBenson) [reply scripts/cnode-helper-scripts/gLiveView.sh:406]: hmmm, strange. 
```suggestion
  clear
```

**Comment by TrevorBenson** [comment]
@asnakep 
Much cleaner. I'll give it a test here shortly and retest the mithril changes as well. :+1: Thanks for all the additions!

**Comment by asnakep** [comment]
Thanks for all the help with git, cheers

File: scripts/cnode-helper-scripts/mithril.library
Comment (TrevorBenson) [inline scripts/cnode-helper-scripts/mithril.library:129]:
```suggestion
  echo "Using RELAY_ENDPOINT=http://${RELAY_ENDPOINT_IP}:${RELAY_PORT} for the Mithril signer relay endpoint."
```

File: scripts/cnode-helper-scripts/gLiveView.sh
Comment (TrevorBenson) [inline scripts/cnode-helper-scripts/gLiveView.sh:786]: I've tested and it looks good and mithril script changes work as intended. I've been contemplating the way to provide support for non Systemd environments. Two thoughts come to mind. 
1. **MITHRIL_SIGNER_TYPE** `(SYSTEMD|OTHER) Type of Mithril Signer Environment`
   * This keeps the current logic where systemd service unit and service status checks are used.
   * When OTHER (anything besides SYSTEMD) is set, only the metrics port responsiveness is used for service status.
   * Unset by default.
2. **MITHRIL_SIGNER_ENABLED** `(Y|N) If the Mithril Signer is enabled`
   * Drops the specific logic for Systemd, and only relies on the metrics server port and responsiveness.
   * Unset by default, but logic defaults to `MITHRIL_SIGNER_ENABLED=N` when variable is not set.

```
mithrilSignerVars() {
  signerMetricsEnabled=$(grep -q "ENABLE_METRICS_SERVER=true" ${CNODE_HOME}/mithril/mithril.env && echo "true" || echo "false")
  echo "singerMetricsEnabled: ${signerMetricsEnabled}" >> /tmp/mithril-gliveview.log
  if [[ "${signerMetricsEnabled}" == "true" ]] ; then
    mithrilSignerMetrics=$(curl -s "http://${METRICS_SERVER_IP}:${METRICS_SERVER_PORT}/metrics" 2>/dev/null | grep -v -E "HELP|TYPE" | sed 's/mithril_signer_//g')
    echo "mithrilSignerMetrics: ${mithrilSignerMetrics}" >> /tmp/mithril-gliveview.log
    if [[ "${MITHRIL_SIGNER_TYPE}" == "SYSTEMD" ]] ; then 
      echo "MITHRIL_SIGNER_TYPE: ${MITHRIL_SIGNER_TYPE}" >> /tmp/mithril-gliveview.log

      signerServiceStatus=$(systemctl status ${CNODE_VNAME}-mithril-signer.service 2>/dev/null | grep 'Active:' | awk '{print $3}' | sed 's/[()]//g')
      echo "signerServiceStatus: ${signerServiceStatus}" >> /tmp/mithril-gliveview.log
    elif [[ -n "${MITHRIL_SIGNER_TYPE}" ]] && [[ ! "${MITHRIL_SIGNER_TYPE}" == "SYSTEMD" ]] ; then
      SIGNER_METRICS_HTTP_RESPONSE=$(curl --write-out "%{http_code}" --silent --output /dev/null --connect-timeout 2 http://${METRICS_SERVER_IP}:${METRICS_SERVER_PORT}/metrics)
      echo "SIGNER_METRICS_HTTP_RESPONSE: ${SIGNER_METRICS_HTTP_RESPONSE}" >> /tmp/mithril-gliveview.log
      if [[ "$SIGNER_METRICS_HTTP_RESPONSE" -eq 200 ]] ; then
        signerServiceStatus='running'
      else
        signerServiceStatus='inactive (dead)'
      fi
      unset SIGNER_METRICS_HTTP_RESPONSE
    fi
  fi
}
```

This change skips trying to curl the metrics server when the metrics are not enabled via the variable. Later the check would drop `serviceExists` an use instead `[[ -n "${MITHRIL_SIGNER_TYPE}" ]]`.

```
      if [[ "$nodemode" = "Core" ]] && [[ -n "${MITHRIL_SIGNER_TYPE}" ]] && [[ "$signerMetricsEnabled" == "true" ]] && [[ "$signerServiceStatus" == "running" ]] ; then
```

Given Systemd and non systemd (manual, containerized, etc.) all rely on the fact that the metrics server has to be used I'm a bit partial to the second solution. The number of branches required in **mithrilSignerVars** would be reduced quite a bit as when the metrics server is not responding the Mithril Signer status could be set to offline, no need to check the systemd service. However, I only provided code suggestion for the first option to keep it aligned with the current PR logic even though it takes more lines/logic to achieve.

@asnakep What are your thoughts?

@rdlrt @Scitz0 Do you have any preference between the two options mentioned? Does a third option come to mind?

Comment (TrevorBenson) [reply scripts/cnode-helper-scripts/gLiveView.sh:786]: This is untested, but I quickly tossed together what I think the function would look like if using **MITHRIL_SIGNER_ENABLED** and not relying on systemd, but simply the metrics server response.

```
mithrilSignerVars() {
  signerMetricsEnabled=$(grep -q "ENABLE_METRICS_SERVER=true" ${CNODE_HOME}/mithril/mithril.env && echo "true" || echo "false")
  if [[ "${MITHRIL_SIGNER_ENABLED}" == "Y" ]] && [[ "${signerMetricsEnabled}" == "true" ]] ; then
    mithrilSignerMetrics=$(curl -s "http://${METRICS_SERVER_IP}:${METRICS_SERVER_PORT}/metrics" 2>/dev/null | grep -v -E "HELP|TYPE" | sed 's/mithril_signer_//g')
    SIGNER_METRICS_HTTP_RESPONSE=$(curl --write-out "%{http_code}" --silent --output /dev/null --connect-timeout 2 http://${METRICS_SERVER_IP}:${METRICS_SERVER_PORT}/metrics)
    if [[ "$SIGNER_METRICS_HTTP_RESPONSE" -eq 200 ]] ; then
      signerServiceStatus='running'
    else
      signerServiceStatus='inactive (dead)'
    fi
    unset SIGNER_METRICS_HTTP_RESPONSE
  fi
}
```

File: scripts/cnode-helper-scripts/gLiveView.sh
Comment (TrevorBenson) [inline scripts/cnode-helper-scripts/gLiveView.sh:1452]: This line prevents gLiveview from showing the Mithril Signer is offline since it requires the state to be running.

File: scripts/cnode-helper-scripts/gLiveView.sh
Comment (rdlrt) [reply scripts/cnode-helper-scripts/gLiveView.sh:786]: Rather than extensive checks every time, regardless of whether mithril signer is being run - I'd prefer to have a variable at the top about enabling Mithril section (defaulting to 'N' ) and then only when the value for that option is not 'Y' , to start looking at options. Thus, your suggestion wrt `MITHRIL_SIGNER_ENABLED` makes sense - just need to ensure this part.

As regards with systemd v/s without, I dont want to put too many combinations in, if enabled and running on a particular port, best to simply use IP:port as a test, exclude systemd completely from code where possible - as it wont be present on docker systems anyways

Comment (asnakep) [reply scripts/cnode-helper-scripts/gLiveView.sh:786]: Hi Trevor and CNTools/Koios Team,

I strongly agree with your additions on check metrics response, mithril signer enabled var and signer type, just please let me know if some action is required from my side, or if this changes will be applied by you.

Thank you
Cheers,
Manuel

Comment (TrevorBenson) [reply scripts/cnode-helper-scripts/gLiveView.sh:786]: I'll make some adjustments and retest for MITHRIL_ENABLED. Submit a new commit later today after some additional testing.

File: scripts/cnode-helper-scripts/gLiveView.sh
Comment (TrevorBenson) [reply scripts/cnode-helper-scripts/gLiveView.sh:1452]: An SPO may decide to copy the required KES hot secret key onto a relay to run a signer decoupled from their Core/Block Producer. They also may have a misconfigured, or previously setup environment missing the variable to enable the metrics server. Since there are multiple variations of how/why the SPO may adjust things compared to our expectations I'm changing this to simply verify that **MITHRIL_SIGNER_ENABLED** is Y. Then the SPO will know if its online, or if offline that something needs to be further investigated.

```suggestion
      if [[ "${MITHRIL_SIGNER_ENABLED}" == "Y" ]] ; then
```

File: scripts/cnode-helper-scripts/gLiveView.sh
Comment (TrevorBenson) [reply scripts/cnode-helper-scripts/gLiveView.sh:786]: Change now gates all steps based on **MITHRIL_SIGNER_ENABLED** being Y, as per @rdlrt suggestion we won't parse files or collect variables unless this is set, even when the metrics server is enabled by the SPO.

This also allows the SPO to run a mithril signer, but decide they want the original gLiveView layout by changing this value to N, without impacting the signer itself or their ability to query its port separately. Would require an update somewhere in documentation to describe it.

```suggestion
mithrilSignerVars() {
  # Require MITHRIL_SIGNER_ENABLED or do not proceed to source files or check ports.
  if [[ "${MITHRIL_SIGNER_ENABLED}" == "Y" ]] ; then
    # mithril.env sourcing needed to have values in ${METRICS_SERVER_IP} and ${METRICS_SERVER_PORT}
    . ${CNODE_HOME}/mithril/mithril.env
    signerMetricsEnabled=$(grep -q "ENABLE_METRICS_SERVER=true" ${CNODE_HOME}/mithril/mithril.env && echo "true" || echo "false")
    if [[ "${signerMetricsEnabled}" == "true" ]] ; then
      mithrilSignerMetrics=$(curl -s "http://${METRICS_SERVER_IP}:${METRICS_SERVER_PORT}/metrics" 2>/dev/null | grep -v -E "HELP|TYPE" | sed 's/mithril_signer_//g')
      SIGNER_METRICS_HTTP_RESPONSE=$(curl --write-out "%{http_code}" --silent --output /dev/null --connect-timeout 2 http://${METRICS_SERVER_IP}:${METRICS_SERVER_PORT}/metrics)
      if [[ "$SIGNER_METRICS_HTTP_RESPONSE" -eq 200 ]] ; then
        signerServiceStatus='online'
      else
        signerServiceStatus='offline'
      fi
      unset SIGNER_METRICS_HTTP_RESPONSE
    fi
  fi
}

```

File: scripts/cnode-helper-scripts/mithril-signer.sh
Comment (TrevorBenson) [inline scripts/cnode-helper-scripts/mithril-signer.sh:230]: Enables gLiveView automatically. We could also inquire with an additional question if they want it enabled. I think when reconfiguring the environment file automatically haveing gLiveView show the signer might be useful to show if a misconfiguration exists and the signer does not start, the SPO can always disable it again in the env file.

```suggestion
    if grep -q "ENABLE_METRICS_SERVER=true" ${CNODE_HOME}/mithril/mithril.env; then
      METRICS_SERVER_PARAMS="--enable-metrics-server --metrics-server-ip ${METRICS_SERVER_IP} --metrics-server-port ${METRICS_SERVER_PORT}"
      # If ENABLE_METRICS_SERVER is true, then an environment update will enable gLiveView automatically.
      sudo sed -i "s/#MITHRIL_SIGNER_ENABLED=[YN]/MITHRIL_SIGNER_ENABLED=Y/" ${CNODE_HOME}/scripts/env
      if ! "${MITHRILBIN}" ${METRICS_SERVER_PARAMS} -vv | tee -a "${LOG_DIR}/$(basename "${0::-3}")".log 2>&1 ; then
```

**Review by TrevorBenson — CHANGES_REQUESTED** [review]
@asnakep I've added comments with code blocks that can be applied in the UI (just remember if you apply in the UI to do a git pull if you want them in your local branch before applying further changes).

Separately the env file requires 1 additional change around [line 113](https://github.com/asnakep/guild-operators/blob/gliveview_mithril-signer/scripts/cnode-helper-scripts/env#L112-L113):

```
#MITHRIL_DOWNLOAD="N"                                   # (Y|N) Download latest Mithril snapshot
#MITHRIL_SIGNER_ENABLED="N"                             # (Y|N) Enable the gLiveView Mithril Signer Section
```

File: scripts/cnode-helper-scripts/gLiveView.sh
Comment (TrevorBenson) [inline scripts/cnode-helper-scripts/gLiveView.sh:1460]: This will resolve the linter/checks in the PR showing red, removing the shellcheck error:

```suggestion
                if [[ $line =~ ${metric_name}[[:space:]]+([0-9]+) ]]; then
```

**Comment by asnakep** [comment]
Thanks so much Trevor for your changes!
Cheers,
Manuel

**Review by TrevorBenson — APPROVED** [review]
Approved. Thanks @asnakep for the great addition.

@rdlrt Since this touches gLiveView.sh I'd appreciate your final review/blessing prior to merge.

**Pull request merged**
